src/time.rs

```diff
@@ -0,0 +1,189 @@
+
+/// A representation of an absolute time value.
+///
+/// The `Instant` type is a wrapper around a `i64` value that
+/// represents a number of milliseconds, monotonically increasing
+/// since an arbitrary moment in time, such as system startup.
+///
+/// * A value of `0` is inherently arbitrary.
+/// * A value less than `0` indicates a time before the starting
+///   point.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Instant {
+    millis: i64,
+}
+
+impl Instant {
+    /// Create a new `Instant` from a number of milliseconds.
+    pub fn from_millis(millis: i64) -> Instant {
+        Instant { millis }
+    }
+
+    /// The fractional number of milliseconds that have passed
+    /// since the beginning of time.
+    pub fn millis(&self) -> i64 {
+        self.millis % MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds that have passed since the
+    /// beginning of time.
+    pub fn secs(&self) -> i64 {
+        self.millis / MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds that have passed since
+    /// the biginning of time.
+    pub fn total_millis(&self) -> i64 {
+        self.millis
+    }
+}
+
+impl fmt::Display for Instant {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}s", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis + rhs.total_millis())
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: Duration) -> Instant {
+        Instant::from_millis(self.millis - rhs.total_millis())
+    }
+}
+
+/// A relative amount of time.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Duration {
+    millis: i64,
+}
+
+impl Duration {
+    /// Create a new `Duration` from a number of milliseconds.
+    pub fn from_millis(millis: i64) -> Duration {
+        Duration { millis }
+    }
+
+    /// The fractional number of milliseconds in this `Duration`.
+    pub fn millis(&self) -> i64 {
+        self.millis % MILLIS_PER_SEC
+    }
+
+    /// The number of whole seconds in this `Duration`.
+    pub fn secs(&self) -> i64 {
+        self.millis / MILLIS_PER_SEC
+    }
+
+    /// The total number of milliseconds in this `Duration`.
+    pub fn total_millis(&self) -> i64 {
+        self.millis
+    }
+}
+
+impl ops::Add<Duration> for Duration {
+    type Output = Duration;
+
+    fn add(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis + rhs.total_millis())
+    }
+}
+
+impl ops::Sub<Duration> for Duration {
+    type Output = Duration;
+
+    fn sub(self, rhs: Duration) -> Duration {
+        Duration::from_millis(self.millis - rhs.total_millis())
+    }
+}
+
+impl ops::Mul<Duration> for Duration {
+    type Output = Duration;
+
+    fn mul(self, rhs: Duration) -> Duration {
```

Comment:
This doesn't make sense... Consider the units. You're multiplying `self` (seconds) by `rhs` (also seconds), so you get seconds squared. But you return seconds!

Thus, `rhs` should be `u32`, just like in `std::time`.

Reply:
Why would the units be seconds here when they're milliseconds everywhere else?

Reply:
I'm talking in terms of [dimensional analysis](https://en.wikipedia.org/wiki/Dimensional_analysis) here. Seconds and milliseconds are the same in the context.